### PR TITLE
Lint on combining `#[no_mangle]` and `#[export_name]`

### LINKED ISSUE #47446: Using both no_mangle and export_name attributes should cause a warning
In a simple example like:

```
#[no_mangle]
#[export_name = "asdf"]
pub extern fn foobar() {}
```

The actual exported name will be `asdf`. It seems `export_name` wins over `no_mangle` however there is no warning letting you know this occurred. It seems there is no valid use for having both of these so it's probably worth warning about.

-B

### PR DESCRIPTION
**Pull request by sassman**
This is my very first contribution to the compiler, even though I read the [chapter about lints](https://rustc-dev-guide.rust-lang.org/diagnostics.html) I'm not very certain that this ~~new lint is done right as a builtin lint~~ PR is right. I appreciate any guidance on how to improve the code.

- Add test for issue #47446 
- ~~Implement the new lint `mixed_export_name_and_no_mangle` as a builtin lint (not sure if that is the right way to go)~~ Extend `unused_attributes` lint
- Add suggestion how to fix it

<details>

<summary>Old proposed new lint</summary>

> The `mixed_export_name_and_no_mangle` lint detects usage of both `#[export_name]` and `#[no_mangle]` on the same item which results on `#[no_mangle]` being ignored.
> 
> *warn-by-default*
> 
> ### Example
> 
> ```rust
> #[no_mangle] // ignored
> #[export_name = "foo"] // takes precedences
> pub fn bar() {}
> ```
> 
> ### Explanation
> 
> The compiler will not respect the `#[no_mangle]` attribute when generating the symbol name for the function, as the `#[export_name]` attribute takes precedence. This can lead to confusion and is unnecessary.

</details>

Fixes rust-lang/rust#47446

### CODE REVIEW

**Comment by workingjubilee** [comment]
@Urgau Do you know why all these lints are prefixed by `Builtin`? Aren't all the lints in the compiler built-in by definition?

**Comment by Urgau** [comment]
> Do you know why all these lints are prefixed by `Builtin`? Aren't all the lints in the compiler built-in by definition?

It's probably a prefix for the lints defined inside [`compiler/rustc_lint/src/builtin.rs`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/builtin.rs).

**Comment by workingjubilee** [comment]
Ah, I see.

Mm, great, in tests/ui/asm/naked-functions.rs we find this gem:
```rust
#[export_name = "exported_function_name"]
#[link_section = ".custom_section"]
#[no_mangle]
#[naked]
pub unsafe extern "C" fn compatible_ffi_attributes_1() {
    naked_asm!("", options(raw));
}
```

@sassman Thank you for PRing this lint! It looks like you will have to fix a few cases of it in our test suite as well, or `--bless`. Feel free to apply either, as you prefer, to relevant cases like that "naked" function.

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (workingjubilee) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:1]:
```suggestion
#![deny(mixed_export_name_and_no_mangle)]
// issue: rust-lang/rust#47446
```

**Comment by sassman** [comment]
Now the tests run all green and I have cleaned up the other test that used both `no_mangle` and `export_name` and simply removed the `no_mangle` in `tests/ui/asm/naked-functions.rs`.

Please let me know if there is anything that I can do further.

**Comment by Urgau** [comment]
I think this check is miss-placed, I think it should be with the others codegen attributes handling in the [`codegen_fn_attrs`](https://github.com/rust-lang/rust/blob/f6648f252a05a0a46c865d7ec836b46290613bf9/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L53C4-L53C20) function in [`compiler/rustc_codegen_ssa/src/codegen_attrs.rs`](https://github.com/rust-lang/rust/blob/f6648f252a05a0a46c865d7ec836b46290613bf9/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L53C4-L53C20).

You can look at the [`check_link_name_xor_ordinal`](https://github.com/rust-lang/rust/blob/f6648f252a05a0a46c865d7ec836b46290613bf9/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L766-L780) function for inspiration: https://github.com/rust-lang/rust/blob/f6648f252a05a0a46c865d7ec836b46290613bf9/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L648

As well as [`TyCtxt::emit_node_span_lint`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.emit_node_span_lint) for emitting the lint.
You will also need to move the lint definition into the `rustc_lint_defs` crate and import that crate in `rustc_codegen_ssa`.

**Comment by sassman** [comment]
@Urgau I have refactored the code, but just to be sure I get things right, with this approach we won't have "named" lints that a user can allow or deny anymore or am I missing something?

In the previous commit the lint was tied to the `LintDiagnostic` struct, that is now not anymore present. 

If I did got this right then I will adjust the introduced error output expectation: 
https://github.com/rust-lang/rust/actions/runs/11332684748/job/31515339659?pr=131558#step:26:3779

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Urgau) [inline compiler/rustc_lint_defs/src/builtin.rs:5109]: Every lint needs to be register to a lint pass, but since this lint emitted outside a lint pass you can add it to the `HardwiredLints` pass at the top of this file:
https://github.com/rust-lang/rust/blob/77348fadcbd3a549556c51f64b4f8f19545e8f14/compiler/rustc_lint_defs/src/builtin.rs#L14-L18

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (Urgau) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:1]: You might also need to do a `build-fail` instead of default `check-fail` since the check  happens very late, i.e. add `//@ build-fail`.

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:821]: As for the `LintDiagnostic` (instead of the manual calls), I think you can add it to the `compiler/rustc_codegen_ssa/src/errors.rs` file of this crate, and use [`emit_node_span_lint`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.emit_node_span_lint) above.

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (sassman) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs]: as it turned out, `check-fail` works but `build-fail` did not work. So I kept the test as it was.

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (sassman) [reply compiler/rustc_codegen_ssa/src/codegen_attrs.rs]: Thanks for pointing me in this direction, I added the first `LintDiagnostic` to the error file, see 71e9a91bae4cd140b2dd01adcafefd70c91fff5e

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (sassman) [reply compiler/rustc_lint_defs/src/builtin.rs:5110]: thanks for pointing out, added it together with the other comments, and all started to work out.

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:808]:
```suggestion
    fn lint_if_mixed(self, tcx: TyCtxt<'_>) {
```

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (Urgau) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:2]: we should check that the suggestion is application and indeed fix the issue
```suggestion
//@ run-rustfix
```

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (Urgau) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:3]: I would prefer for the test to use the default lint level, so if it changes it's reflected in the test.
```suggestion
//@ check-pass
```

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (Urgau) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:7]:
```suggestion
//~^ WARN the attribute `export_name` may not be used in combination with `no_mangle`
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Urgau) [inline compiler/rustc_lint_defs/src/builtin.rs:5089]:
```suggestion
    /// The `mixed_export_name_and_no_mangle` lint detects usage of both
    /// `#[export_name]` and `#[no_mangle]` on the same item which result
    /// on `#[no_mangle]` being ignored.
```

File: compiler/rustc_codegen_ssa/src/errors.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/errors.rs:1098]: We are no longer in the `builtin.rs` file, let's remove the prefix.
```suggestion
pub(crate) struct MixedExportNameAndNoMangle {
```

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:787]: Let's reduce a bit the length of this struct.
```suggestion
struct MixedExportNameAndNoMangleState {
```

File: compiler/rustc_codegen_ssa/messages.ftl
Comment (Urgau) [inline compiler/rustc_codegen_ssa/messages.ftl:193]: attributes should always be displayed with their `#[..]` form 
```suggestion
codegen_ssa_mixed_export_name_and_no_mangle = the `#[export_name]` attribute may not be used in combination with `#[no_mangle]`
```

File: compiler/rustc_codegen_ssa/messages.ftl
Comment (Urgau) [inline compiler/rustc_codegen_ssa/messages.ftl:194]:
```suggestion
    .label = `#[export_name]` takes precedence
```

File: compiler/rustc_codegen_ssa/messages.ftl
Comment (Urgau) [inline compiler/rustc_codegen_ssa/messages.ftl:195]:
```suggestion
    .note = `#[no_mangle]` is ignored
```

File: compiler/rustc_codegen_ssa/messages.ftl
Comment (Urgau) [inline compiler/rustc_codegen_ssa/messages.ftl:196]:
```suggestion
    .suggestion = remove the `#[no_mangle]` attribute
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Urgau) [inline compiler/rustc_lint_defs/src/builtin.rs:5112]:
```suggestion
    "mixed usage of `#[export_name]` and `#[no_mangle]`, where `#[export_name]` takes precedence"
```

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:796]: nit: `#[derive(Default)]` is less code

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:810]: nit: in this case we want to be exhaustive, we could use a de-structuring pattern to ensure that
```suggestion
        if let Self {
            export_name: Some(export_name),
            no_mangle: Some(no_mangle),
            hir_id: Some(hir_id),
        } = self
```

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:804]:
```suggestion
        self.no_mangle = Some(span);
```

**Comment by Urgau** [comment]
Reminder, don't forget to use [the `rustbot` review commands](https://github.com/rust-lang/rust/pull/131558#issuecomment-2407749951).

**Comment by davidtwco** [comment]
@Urgau is happy with the state of this PR so triggering an FCP since that's what there is precedent for when adding a new lint.

cc @rust-lang/lang for visibility, this seems within t-compiler's remit but just in case you disagree

@rfcbot fcp merge

**Comment by sassman** [comment]
@nagisa symbol alias sounds interesting, then also multiple `#[export_name]` attributes must be supported, and `#[no_mangle]` would become a shortcut just as you described it.

Is there already an RFC or any tracking issue for this?

**Comment by sassman** [comment]
I guess I found the pre-RFC: https://internals.rust-lang.org/t/pre-rfc-defining-function-aliases/11424

**Comment by sassman** [comment]
The unused attribute lint sounds also reasonable.

Would it then be possible to indicate how to fix it and elaborate on the details as well?

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Urgau) [inline compiler/rustc_lint_defs/src/builtin.rs:5114]:
```suggestion
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Urgau) [inline compiler/rustc_lint_defs/src/builtin.rs:70]:
```suggestion
```

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:813]:
```suggestion
                lint::builtin::UNUSED_ATTRIBUTES,
```

File: tests/ui/attributes/mixed_export_name_and_no_mangle.stderr
Comment (Urgau) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.stderr:12]:
```suggestion
   = note: `#[warn(unused_attributes)]` on by default
```

Comment (sassman) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.stderr:12]: With the `unused_attributes` now in use, would it also make sense to switch the "unused" attribute to the primary span and make the `export_name` attribute then the secondary? So in fact just switch the 2 attributes in the error message.

My idea was also that it would make most sense that the  `#[allow(unused_attributes)]` then is put to the `#[no_mangle]` attribute and not to the `#[export_name]` attribute.

Would you agree on that?

Comment (Urgau) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.stderr:12]: Yeah, that makes sense. Let's put the warning on the `#[no_mangle]`.

Comment (sassman) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.stderr]: implemented with 338f8f73da8b39d19ce4f3e527a076ef79857005

**Comment by sassman** [comment]
I refactored according to the suggestions, but now the test fails as if the lint is not emitted at all anymore. Probably I just miss something, on the registration part.

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (Urgau) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:4]: `compiletest` automatically [adds `-A unused`](https://github.com/rust-lang/rust/blob/1350eead10c46b9d3c2007fe0ea7892b7d7337ab/src/tools/compiletest/src/runtest.rs#L1603) for convenience, here we need to override this 
```suggestion

#![warn(unused_attributes)]

```

**Comment by CryZe** [comment]
Does this work with the new unsafe attributes as well? [For reference](https://blog.rust-lang.org/2024/10/17/Rust-1.82.0.html#unsafe-attributes)

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (sassman) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs:4]: Thanks a lot, it would have taken me like forever to figure that out.

**Comment by sassman** [comment]
@CryZe good question, I'm going to add a test for it, but I suspect it doesn't.

**Comment by sassman** [comment]
@CryZe it works for `#[unsafe(no_mangle)]` but, 

The error message I would want should be this: 
```
warning: the `#[unsafe(no_mangle)]` attribute may not be used in combination with `#[export_name]`
```

But instead the attribute does not have the `unsafe()` surrounding, so it is 

```
warning: the `#[no_mangle]` attribute may not be used in combination with `#[export_name]`
```

But with showing the spans this might be somehow confusing.
Where is the `unsafe` attribute handled anyways?

File: tests/ui/attributes/mixed_export_name_and_no_mangle.rs
Comment (sassman) [inline tests/ui/attributes/mixed_export_name_and_no_mangle.rs:11]: @Urgau maybe you have an idea why this `unsafe` wrapper is not showing up in the error message?

Comment (Urgau) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs:11]: That's because you are using `to_ident_string()` which only look at the identifier but an attribute is composed of many more things than a single identifier.

For good reproduction I would suggest using the AST pretty printer (crate `rustc_ast_pretty`), [rustc_ast_pretty::pprust::attribute_to_string](https://doc.rust-lang.org/stable/nightly-rustc/rustc_ast_pretty/pprust/fn.attribute_to_string.html), but looking at the code I don't thing it handles safety yet (which is a bug).

I would suggest leaving for now the safety part out and having a `// FIXME: Handle attribute unsafety` somewhere in the code.

Alternatively you could already add the call to the pretty printer, add the fixme, and if you feel like it fix the pretty-printer in another PR.

But in anycase, let's not block the PR on this, let's make CI pass, so we can resolve the concern.

Comment (sassman) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs:11]: Thanks for your suggestion, I've used now the pretty printer, the good thing it prints already the full attributes with the `#[]` surroundings,

But unfortunately the `unsafe()` encapsulation is still not showing up.

Comment (Urgau) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs:11]: It's probably around here, https://github.com/rust-lang/rust/blob/3ec4308f6cb4bba3140d736d9ebd53b234fa7df8/compiler/rustc_ast_pretty/src/pprust/state.rs#L628-L630, if you want you can try fixing it (in another pr), but in the mean time let's move on this PR with a `FIXME`, so we can unblock the concern.

Comment (Urgau) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs:11]: Opened #132180 to fix the pretty-printing bug.

Comment (sassman) [reply tests/ui/attributes/mixed_export_name_and_no_mangle.rs:11]: @Urgau, thanks to your fix, the error message now adds up to how it should. So, I guess everything is addressed and solved now.

File: compiler/rustc_codegen_ssa/Cargo.toml
Comment (Urgau) [inline compiler/rustc_codegen_ssa/Cargo.toml:68]: Unnecessary change, revert this change

File: compiler/rustc_codegen_ssa/Cargo.toml
Comment (Urgau) [inline compiler/rustc_codegen_ssa/Cargo.toml:72]: ...same with this one

File: compiler/rustc_codegen_ssa/messages.ftl
Comment (Urgau) [inline compiler/rustc_codegen_ssa/messages.ftl:193]:
```suggestion
codegen_ssa_mixed_export_name_and_no_mangle = `{$no_mangle_attr_name}` attribute may not be used in combination with `#[export_name]`
```

File: compiler/rustc_codegen_ssa/src/errors.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/errors.rs:1101]:
```suggestion
    pub no_mangle_attr: String,
```

Comment (sassman) [reply compiler/rustc_codegen_ssa/src/errors.rs]: done

File: compiler/rustc_codegen_ssa/Cargo.toml
Comment (sassman) [reply compiler/rustc_codegen_ssa/Cargo.toml]: done

**Comment by Urgau** [comment]
@petrochenkov [your concern](https://github.com/rust-lang/rust/pull/131558#issuecomment-2420097066) about the unjustified new lint has been resolved by @sassman *(the PR author)* by issuing the `unused_attributes` lint instead as asked.

The concern can (I think) be marked as resolved.

As a procedural note, with the removal of the new lint, does the FCP still makes sense?
Feels a bit heavy, now that it's such a small addition.

**Comment by sassman** [comment]
Is there any further thing I can do to support the process?

**Comment by jackh726** [comment]
I agree this no longer needs an FCP. Maybe @davidtwco or another person on the FCP can weigh in to agree or dissent?

**Comment by sassman** [comment]
@Urgau is there anything that I can do to support on landing this PR?

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:795]: We should try store a reference to the attribute instead of pretty-printing it every time.
```suggestion
    no_mangle_attr: Option<&'a ast::Attribute>,
```

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:123]:
```suggestion
                        attr,
```

File: compiler/rustc_codegen_ssa/src/codegen_attrs.rs
Comment (Urgau) [inline compiler/rustc_codegen_ssa/src/codegen_attrs.rs:824]:
```suggestion
                    no_mangle_attr: rustc_ast_pretty::pprust::attribute_to_string(no_mangle_attr),
```

**Review by Urgau — APPROVED** [review]

**Comment by Kobzol** [comment]
@rust-timer build 79254c61b038a0f5cb2d08342d00b7057f9db12a

**Pull request merged**
